### ISSUE TITLE
Warn before closing a form with unsaved changes

### ISSUE THREAD

**Issue by taylordowns2000** [OpenFn/Lightning#780]
Starting with the job builder because it's the most sensitive, we need an elegant and generic way to detect when a form has unsaved changes (maybe through the validate hook?) then then display a warning before the user closes.

This is particularly important on the job builder since clicking anywhere _else_ on the canvas will close the form.

**Comment by NickOpenFn** [comment]
Do in the next sprint, requires UI

**Comment by NickOpenFn** [comment]
Noting that this is badly spec'd for taking into the sprint

**Comment by elias-ba** [comment]
Technically blocked. See [this question](https://elixirforum.com/t/help-needed-preventing-liveview-navigation-when-form-is-dirty/57895?u=elias) on the Elixir Forum.

**Comment by NickOpenFn** [comment]
Moving into backlog. When planned again will need a spike to look at alternatives.

**Comment by taylordowns2000** [comment]
@stuartc , this also realtes to #2007 ?

**Comment by christad92** [comment]
@midigofrank's work on #2297 should help with this. Modals will not close by clicking outside them.

@midigofrank correct?

**Comment by midigofrank** [comment]
@christad92 yes and no. I don't think the modals in the workflow page uses the same component as the one in #2297. 
But we can use the same technique.

**Comment by taylordowns2000** [comment]
Yeah for sure. Something we might be better able to handle with React @doc-han ?

**Comment by doc-han** [comment]
> Yeah for sure. Something we might be better able to handle with React [@doc-han](https://github.com/doc-han) ?

That can be handled perfectly if we introduce local persistence to the store.

**Comment by doc-han** [comment]
thinking about it, persisting the entire store might be tricky. Because our store is currently live. 
changes pushed to the store automatically get sent to phx and then phx sends back a patch for us to apply. 

Hence when there's a change. 
- react store upon getting the change pushes it to phx
- phx applied the change to it's storage object and send back a patch to the react store which we apply. (note: phx hasn't committed to db yet)
- upon save clicked, we commit to db. 


**To persist changes locally**. 
1. take a snapshot of the store(locally on their machine) at the point a user visits the editor and get `current_workflow_params`
2. persist all user changes (locally on their machine). Hence, they can leave the page anytime. 
3. Upon return, if the new `current_workflow_params` has a diff from 1. it means updates were done somewhere else. either by someone or on a different machine. we'll have to notify the user.
... several routes

**Comment by doc-han** [comment]
Currently, all user changes are sent to backend and is kept in the livesocket `@assigns`. as soon as the connection is lost. that data is also lost.

**Comment by theroinaochieng** [comment]
Moving this back to the product backlog for reprioritization.

**Comment by taylordowns2000** [comment]
@doc-han , after #4289 goes in, would we have the ability to show a simple browser alert when navigating away from the collaborative editor?

<img width="412" height="268" alt="Image" src="https://github.com/user-attachments/assets/dbb2b377-5695-47bd-9489-b154ca66cd40" />

Because changes are _kept_ in the y-doc, the language should probably say: "You have unsaved changes. Would you like to save before leaving?"

**Comment by doc-han** [comment]
> [@doc-han](https://github.com/doc-han) , after [#4289](https://github.com/OpenFn/lightning/pull/4289) goes in, would we have the ability to show a simple browser alert when navigating away from the collaborative editor?
> 
> <img alt="Image" width="412" height="268" src="https://private-user-images.githubusercontent.com/8732845/533347349-dbb2b377-5695-47bd-9489-b154ca66cd40.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc4Nzk3NTEsIm5iZiI6MTc2Nzg3OTQ1MSwicGF0aCI6Ii84NzMyODQ1LzUzMzM0NzM0OS1kYmIyYjM3Ny01Njk1LTQ3YmQtOTQ4OS1iMTU0Y2E2NmNkNDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDEwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAxMDhUMTMzNzMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGYwZjc0YTczNGE1NDlhNmZlZmQxYWMxN2NlZGRiNTM3YTBhYWE1NjNiNTg3YmUwYmQyNWQxNTc3YzAwNDQyMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.6drxZmNAsn89-o7wHaxUYktXLiCj6rrIt6hW9wuWuUI">
> Because changes are _kept_ in the y-doc, the language should probably say: "You have unsaved changes. Would you like to save before leaving?"

I doubt the browsers give us control on what the wording should be.
